nemo_aligner/algorithms/generation.py

```diff
@@ -0,0 +1,508 @@
+                prompt_lens = global_batch["prompt_lens"]
+                gen_lens = global_batch["gen_lens"]
+                valids = global_batch["valids"]
+
+                gen_tokens_list = [torch.zeros_like(gen_tokens) for _ in range(dp_group.size())]
+                prompt_lens_list = [torch.zeros_like(prompt_lens) for _ in range(dp_group.size())]
+                gen_lens_list = [torch.zeros_like(gen_lens) for _ in range(dp_group.size())]
+                valids_list = [torch.zeros_like(valids) for _ in range(dp_group.size())]
+
+                torch.distributed.all_gather(gen_tokens_list, gen_tokens, group=dp_group)
+                torch.distributed.all_gather(prompt_lens_list, prompt_lens, group=dp_group)
+                torch.distributed.all_gather(gen_lens_list, gen_lens, group=dp_group)
+                torch.distributed.all_gather(valids_list, valids, group=dp_group)
+
+                self.consumed_samples += self.model.cfg.global_batch_size
+                self.step += 1
+
+                if torch.distributed.get_rank() == 0 and gen_tokens_list is not None:
+                    for t, s, e, v in zip(gen_tokens_list, prompt_lens_list, gen_lens_list, valids_list):
+                        buffer = [[] for _ in range(t.shape[1])]
+                        for idx in range(len(t)):
+                            for pdx, (t_, s_, e_, v_) in enumerate(
+                                zip(t[idx], s[idx].tolist(), e[idx].tolist(), v[idx].tolist())
+                            ):
+                                prompt = self.model.tokenizer.ids_to_text(t_[:s_].long().tolist())
+                                response = self.model.tokenizer.ids_to_text(t_[s_:e_].long().tolist())
+                                if v_:
+                                    buffer[pdx].append((prompt, response))
+
+                        for cand_list in buffer:
+                            if len(cand_list) == 0:
+                                continue
+                            assert all([cand_list[0][0] == x[0] for x in cand_list]), "all prompts in group not equal"
+                            payload = {
+                                "step": self.step,
+                                "consumed_samples": self.consumed_samples,
+                                "prompt": cand_list[0][0],
+                                "responses": list(set([x[1] for x in cand_list])),
+                            }
+                            self.generations_fh.write(json.dumps(payload, ensure_ascii=False) + "\n")
+                torch.distributed.barrier()
+
+                run_time_exceeded = self.run_timer.is_finished()
+                if run_time_exceeded:
+                    logging.info(f"Time limit given by run_timer={self.run_timer} reached. Stopping run")
+                    return
+
+        self.logger.finalize()
+
+        if torch.distributed.get_rank() == 0 and self.generations_fh is not None:
```

Comment:
Should never be None, right?
```suggestion
        if torch.distributed.get_rank() == 0:
```